### DISCUSSION TITLE
Backend login using SSO

### DISCUSSION THREAD

**Discussion by michaelclaesson** [pimcore/pimcore#9204] [Q&A / Support] [answered]
Hi!

I'm wondering whether it's possible to override the regular login behavior to allow for single sign on using for instance Microsoft 365?

We have a client interested in setting up a solution like this, but no idea whether or not it is possible. This would work by connecting a Pimcore user to a Azure user ID. So given a valid user from Azure we would immediately login the corresponding Pimcore user without having to provide the password. 

Cheers,
Michael

**Comment by passioneight** [comment]
Hi Michael,

yes you can do this. However, currently there's no nice way to do this. 

You'll have to high-jack the response and split the HTML e.g. using `</form>` . Then add your custom HTML either before or after the form and glue the HTML back together, correspondingly.

[Here's a loosely related discussion](https://github.com/pimcore/pimcore/discussions/8792).

**Reply by michaelclaesson** [reply]
Oh that's unfortunate. Thanks! And what about the back end side of things. Any documentation? How would I authenticate the user? I'd rather not dive into it without any clear idea of what I'm up against.

**Reply by passioneight** [reply]
Well, I'm guessing you want to use something like `Okta` or `OneLogin`, correct?

Try their corresponding SAML 2.0 libraries, if they have any - e.g., [onelogin/php-saml](https://github.com/onelogin/php-saml).

You'll need a controller to trigger the SSO once the user clicks on your button/link and handle the response of the `IdP` (Identity Provider).

> You'll be the `SP` (Service Provider).

When receiving a success response, you'll want to manually create a Pimcore user and log them in.

**Reply by aarongerig** [reply]
@michaelclaesson FYI here's my feature request for exactly that. https://github.com/pimcore/pimcore/issues/8510

**Reply by michaelclaesson** [reply]
> Well, I'm guessing you want to use something like `Okta` or `OneLogin`, correct?
> 
> Try their corresponding SAML 2.0 libraries, if they have any - e.g., [onelogin/php-saml](https://github.com/onelogin/php-saml).
> 
> You'll need a controller to trigger the SSO once the user clicks on your button/link and handle the response of the `IdP` (Identity Provider).
> 
> > You'll be the `SP` (Service Provider).
> 
> When receiving a success response, you'll want to manually create a Pimcore user and log them in.

Correct, but in our case we would use Microsoft 365 and Azure AD identities. The actual communication with Microsoft/Azure is no issue. We have used this frontend to get name and such with success. What I'm unsure about is how I would go about to login the corresponding Pimcore user if any found since I can't find any documentation on this.

There's a firewall and guard in place and I'm not sure how I would manually authenticate a user with code in my custom controller so to speak. Hope I'm making sense. 🙂 In that case hijacking the regular login would not be an issue because we could just provide an alternative SSO login URL.

**Reply by michaelclaesson** [reply]
Can I still use the Authentication tool provided by Pimcore and login using the token redirect method?

https://talk.pimcore.org/t/single-sign-on-using-okta-in-pimcore-adminpanel/4001

**Reply by passioneight** [reply]
Well, I haven't tried it, but you could just give it a go - simply create an action that creates/loads a Pimcore user and then redirects to the backend. 

For anyone else reading this on their phone, these are the corresponding lines necessary (link to the soruce code was provided by @michaelclaesson):
```php
//login as user
$user = User::getByName($isid);
$token = Tool\Authentication::generateToken($user->getName(), $user->getPassword());
$link = $request->getScheme() . '://' . $request->getHttpHost() . '/admin/login/login?username=' . $user->getName() . '&token=' . $token;
return $this->redirect($link);
```

**Reply by michaelclaesson** [reply]
I just wanted to confirm that this works!

**Comment by michaelclaesson** [comment, marked as answer]
This has been tested and solved using the following code.

```php
//login as user
$user = User::getByName($isid);
$token = Tool\Authentication::generateToken($user->getName(), $user->getPassword());
$link = $request->getScheme() . '://' . $request->getHttpHost() . '/admin/login/login?username=' . $user->getName() . '&token=' . $token;
return $this->redirect($link);
```
